// emery/dist/emery.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var assertions_dist_emeryAssertions = require('../assertions/dist/emery-assertions.cjs.dev.js');
var number = require('./number-1c017a79.cjs.dev.js');
var guards_dist_emeryGuards = require('../guards/dist/emery-guards.cjs.dev.js');
var opaques_dist_emeryOpaques = require('../opaques/dist/emery-opaques.cjs.dev.js');
var object = require('./object-e58860b1.cjs.dev.js');



exports.assert = assertions_dist_emeryAssertions.assert;
exports.assertNever = assertions_dist_emeryAssertions.assertNever;
exports.warning = assertions_dist_emeryAssertions.warning;
exports.checkAll = number.checkAll;
exports.checkAllWith = number.checkAllWith;
exports.isEven = number.isEven;
exports.isFinite = number.isFinite;
exports.isFloat = number.isFloat;
exports.isInfinite = number.isInfinite;
exports.isInteger = number.isInteger;
exports.isNegative = number.isNegative;
exports.isNegativeZero = number.isNegativeZero;
exports.isNonNegative = number.isNonNegative;
exports.isNonPositive = number.isNonPositive;
exports.isOdd = number.isOdd;
exports.isPositive = number.isPositive;
exports.negate = number.negate;
exports.isBoolean = guards_dist_emeryGuards.isBoolean;
exports.isDefined = guards_dist_emeryGuards.isDefined;
exports.isFulfilled = guards_dist_emeryGuards.isFulfilled;
exports.isNonEmptyArray = guards_dist_emeryGuards.isNonEmptyArray;
exports.isNull = guards_dist_emeryGuards.isNull;
exports.isNullish = guards_dist_emeryGuards.isNullish;
exports.isNumber = guards_dist_emeryGuards.isNumber;
exports.isRejected = guards_dist_emeryGuards.isRejected;
exports.isString = guards_dist_emeryGuards.isString;
exports.isUndefined = guards_dist_emeryGuards.isUndefined;
exports.castToOpaque = opaques_dist_emeryOpaques.castToOpaque;
exports.getErrorMessage = object.getErrorMessage;
exports.typedEntries = object.typedEntries;
exports.typedKeys = object.typedKeys;
